nbconverted/2_tsg_analysis.py

```diff
@@ -0,0 +1,317 @@
+plt.title(gene)
+
+
+# In[18]:
+
+
+# merge predictions with sample mutation count info
+probs_df = preds_df.merge(sample_count, how='left', left_index=True, right_index=True)
+probs_df.rename(columns={'Hugo_Symbol': 'mutation_count'}, inplace=True)
+probs_df['mutation_count'] = probs_df['mutation_count'].fillna(0).astype(int)
+
+# threshold mutation count into 3 categories (0/1/2+ total mutations)
+probs_df.loc[probs_df.mutation_count > 1, 'mutation_count'] = '2+'
+probs_df['mutation_count'] = probs_df.mutation_count.astype(str)
+
+probs_df.head()
+
+
+# In[19]:
+
+
+# make sure that all samples with a true class of 1 have 1 or more mutations observed
+# and make sure that all samples with a true class of 0 have 0 mutations observed
+#
+# this is essentially just a check that our downloaded maf file matches what was
+# used to generate the classifier labels (it should)
+assert np.array_equal(
+    probs_df.loc[probs_df.true_class == 0, :].mutation_count.unique(), ['0']
+)
+assert np.array_equal(
+    probs_df.loc[probs_df.true_class == 1, :].mutation_count.unique(), ['1', '2+']
+)
+
+
+# In[20]:
+
+
+sns.set({'figure.figsize': (12, 6)})
+sns.violinplot(data=probs_df, x='mutation_count', y='positive_prob')
+plt.xlabel('Number of mutations observed')
+plt.ylabel('Predicted probability of mutation')
+plt.title('{} observed mutation count vs. predicted probability, averaged over all cancer types'.format(gene))
+
+
+# ### Segment analysis by cancer type
+# 
+# Do the same as above, but with samples split by cancer type. We'll display the top few cancer types (by number of samples) for each variant.
+
+# In[21]:
+
+
+exp_sample_info_file = Path('/home/jake/research/mpmp/data/sample_info/tcga_expression_sample_identifiers.tsv')
+sample_info_df = pd.read_csv(exp_sample_info_file, sep='\t', index_col=0)
+sample_info_df.head()
+
+
+# In[22]:
+
+
+probs_cancer_type_df = (probs_df
+    .merge(
+        sample_info_df.drop(columns=['sample_type', 'id_for_stratification']),
+        left_index=True, right_index=True
+    )
+)
+probs_cancer_type_df.head()
+
+
+# In[23]:
+
+
+cancer_type_counts = (probs_cancer_type_df
+    .groupby('cancer_type')
+    .count()
+    .fold_no
+    .sort_values(ascending=False)
+)
+cancer_type_counts.head()
+
+
+# In[24]:
+
+
+sns.set({'figure.figsize': (24, 14)})
+fig, axarr = plt.subplots(2, 3)
+
+for ix, cancer_type in enumerate(cancer_type_counts.index[:6]):
+    print(cancer_type)
+    ax = axarr[ix // 3, ix % 3]
+    plot_df = (
+        probs_cancer_type_df[probs_cancer_type_df.cancer_type == cancer_type]
+          .sort_values(by='mutation_count')
+    )
+    sns.violinplot(data=plot_df, x='mutation_count', y='positive_prob',
+                   inner='point', dodge=True, cut=0, ax=ax)
+    ax.set_title(gene + ', ' + cancer_type)
+    ax.set_xlabel('Number of mutations observed')
+    ax.set_ylabel('Predicted probability of mutation')
+    ax.set_ylim(-0.1, 1.1)
+    
+plt.tight_layout()
```

Comment:
Interesting that the APC results look so good across all cancer types and then much less significant in individual cancer types. Is that just a sample size issue, do you think?

Reply:
Good question! I don't think so - I think this is specific to APC, and it's probably not a great example of sample-specific effects to use going forward. The 2+ mutation samples seem to be dominated by colon cancer (COAD), where APC inactivation is [an initiating event in carcinogenesis](https://pubmed.ncbi.nlm.nih.gov/11978510/). My guess, which I haven't looked into yet but may in the future, is that a lot of the 1-mutation samples in COAD are effectively 2-mutation samples through some other mechanism (epigenetic silencing, homozygous copy deletion which we're not capturing here, etc) since APC inactivation is basically a requirement for colon cancer development.

The two "modes" in the 1-mutation distribution are probably just COAD samples (high prediction of APC mutation) and other cancer types (lower prediction of APC mutation). In that way, it does seem to be a good cautionary example about the dangers of looking at averages across cancer types.

For other genes I haven't noticed as strong of a segmentation by cancer type, so I think it's more likely that the bimodality/high variance in the single mutation samples is actually saying something about those samples. We will see!